src/lsm/tree.zig

```diff
@@ -693,155 +687,112 @@ pub fn TreeType(comptime Table: type, comptime Storage: type, comptime tree_name
             assert(tree.compaction_callback != null);
 
-            // compact_done() is called after all compact_io_tick()'s complete.
+            // compact_done() is called after all compact_tick()'s complete.
             // This function can be triggered asynchronously or by compact_cpu() below.
             tree.compaction_io_pending -= 1;
             if (tree.compaction_io_pending == 0) tree.compact_done();
         }
 
-        fn compact_cpu(tree: *Tree) void {
-            assert(tree.compaction_io_pending <= 2 + tree.compaction_table.len);
-            assert(tree.compaction_callback != null);
-
-            // Try to tick the cpu portion of the immutable table compaction:
-            const compaction_beat = tree.compaction_op % config.lsm_batch_multiple;
-            const even_levels = compaction_beat < half_measure_beat_count;
-            if (even_levels) {
-                assert(tree.compaction_table_immutable.status == .idle);
-            } else {
-                if (tree.compaction_table_immutable.status == .compacting) {
-                    tree.compaction_table_immutable.tick_cpu();
-                }
-            }
-
-            // Try to tick the cpu portion of the level compactions:
-            var it = CompactionTableIterator{ .tree = tree };
-            while (it.next()) |context| {
-                if (context.compaction.status == .compacting) {
-                    assert(context.compaction.level_b == context.level_b);
-                    context.compaction.tick_cpu();
-                }
-            }
-
-            // Resolve the io_pending added by compact_io(). This may trigger compact_done().
-            tree.compact_io_tick_done();
-        }
-
         fn compact_done(tree: *Tree) void {
             assert(tree.compaction_io_pending == 0);
             assert(tree.compaction_callback != null);
 
+            // Will be true if there's any Compactions still processing.
             var still_compacting = false;
 
-            // Mark immutable compaction that reported done in their callback as "completed".
+            // Reset the immutable table Compaction once it finishes.
+            // Also clear any invisible tables it created during compaction.
             const compaction_beat = tree.compaction_op % config.lsm_batch_multiple;
             const even_levels = compaction_beat < half_measure_beat_count;
             if (even_levels) {
                 assert(tree.compaction_table_immutable.status == .idle);
             } else {
-                if (tree.compaction_table_immutable.status == .done) {
-                    tree.compaction_table_immutable.reset();
-                    tree.table_immutable.clear();
-                } else if (tree.compaction_table_immutable.status == .compacting) {
-                    still_compacting = true;
+                switch (tree.compaction_table_immutable.status) {
+                    .idle => {}, // wasn't started for this half measure
```

Comment:
If we didn't start it this half measure, I assume the immutable table is empty? Can we assert `tree.compaction_table_immutable.free`?